File: packages/server/src/routes/reactions.ts
import type { FastifyInstance } from 'fastify';
import { getDb } from '../db.js';
import * as Q from '../queries.js';
import { broadcastToChannel } from '../ws.js';

export function registerReactionRoutes(app: FastifyInstance): void {
  app.put<{
    Params: { messageId: string };
    Body: { emoji: string };
  }>('/api/v1/messages/:messageId/reactions', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const { messageId } = request.params;
    const { emoji } = request.body ?? {};
    if (!emoji || typeof emoji !== 'string') {
      return reply.status(400).send({ error: 'emoji is required' });
    }

    const db = getDb();
    const message = Q.getMessageById(db, messageId);
    if (!message) return reply.status(404).send({ error: 'Message not found' });

    if (!Q.isChannelMember(db, message.channel_id, userId)) {
      return reply.status(403).send({ error: 'Not a member of this channel' });
    }

    const distinctCount = Q.getReactionCountForMessage(db, messageId);
    if (distinctCount >= 20) {
      const existing = db.prepare(
        'SELECT 1 FROM message_reactions WHERE message_id = ? AND emoji = ? LIMIT 1',
      ).get(messageId, emoji);
      if (!existing) {
        return reply.status(429).send({ error: 'Maximum 20 different emoji reactions per message' });
      }
    }

    Q.addReaction(db, messageId, userId, emoji);
    const reactions = Q.getReactionsByMessageId(db, messageId);
    broadcastToChannel(message.channel_id, {
      type: 'reaction_update',
      message_id: messageId,
      channel_id: message.channel_id,
      reactions,
    });
    return { ok: true, reactions };
  });

  app.delete<{
    Params: { messageId: string };
    Body: { emoji: string };
  }>('/api/v1/messages/:messageId/reactions', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const { messageId } = request.params;
    const { emoji } = request.body ?? {};
    if (!emoji || typeof emoji !== 'string') {
      return reply.status(400).send({ error: 'emoji is required' });
    }

    const db = getDb();
    const message = Q.getMessageById(db, messageId);
    if (!message) return reply.status(404).send({ error: 'Message not found' });

    if (!Q.isChannelMember(db, message.channel_id, userId)) {
      return reply.status(403).send({ error: 'Not a member of this channel' });
    }

    Q.removeReaction(db, messageId, userId, emoji);
    const reactions = Q.getReactionsByMessageId(db, messageId);
    broadcastToChannel(message.channel_id, {
      type: 'reaction_update',
      message_id: messageId,
      channel_id: message.channel_id,
      reactions,
    });
    return { ok: true, reactions };
  });

  app.get<{
    Params: { messageId: string };
  }>('/api/v1/messages/:messageId/reactions', async (request, reply) => {
    const userId = request.currentUser?.id;
    if (!userId) return reply.status(401).send({ error: 'Authentication required' });

    const { messageId } = request.params;
    const db = getDb();
    const message = Q.getMessageById(db, messageId);
    if (!message) return reply.status(404).send({ error: 'Message not found' });

    if (!Q.isChannelMember(db, message.channel_id, userId)) {
      return reply.status(403).send({ error: 'Not a member of this channel' });
    }

    const reactions = Q.getReactionsByMessageId(db, messageId);
    return { reactions };
  });
}

File: packages/server-go/internal/api/reactions.go
package api

import (
	"log/slog"
	"net/http"

	"borgee-server/internal/store"
)

type ReactionHandler struct {
	Store  *store.Store
	Logger *slog.Logger
	Hub    EventBroadcaster
}

func (h *ReactionHandler) RegisterRoutes(mux *http.ServeMux, authMw func(http.Handler) http.Handler) {
	wrap := func(f http.HandlerFunc) http.Handler { return authMw(f) }

	mux.Handle("PUT /api/v1/messages/{messageId}/reactions", wrap(h.handleAddReaction))
	mux.Handle("DELETE /api/v1/messages/{messageId}/reactions", wrap(h.handleRemoveReaction))
	mux.Handle("GET /api/v1/messages/{messageId}/reactions", wrap(h.handleGetReactions))
}

// AP-4 立场 ①+③ — channel-member ACL gate (CV-7 #535 既存 gap 闭合).
// Reuses Store.IsChannelMember + Store.CanAccessChannel (跟 messages.go
// 既有 ACL 同源). Returns true iff caller may operate on reactions of
// `messageID`. On false, caller MUST emit 404 "Channel not found"
// byte-identical to messages.go::handleCreateMessage line 230-232 (channel
// hidden from non-member, fail-closed).
func (h *ReactionHandler) canAccessMessage(user *store.User, messageID string) (*store.Message, bool) {
	msg, err := h.Store.GetMessageByID(messageID)
	if err != nil {
		return nil, false
	}
	if !h.Store.IsChannelMember(msg.ChannelID, user.ID) || !h.Store.CanAccessChannel(msg.ChannelID, user.ID) {
		return nil, false
	}
	return msg, true
}

func (h *ReactionHandler) handleAddReaction(w http.ResponseWriter, r *http.Request) {
	user, ok := mustUser(w, r)
	if !ok {
		return
	}

	messageID := r.PathValue("messageId")
	msg, ok := h.canAccessMessage(user, messageID)
	if !ok {
		writeJSONError(w, http.StatusNotFound, "Channel not found")
		return
	}

	var body struct {
		Emoji string `json:"emoji"`
	}
	if err := readJSON(r, &body); err != nil {
		writeJSONError(w, http.StatusBadRequest, err.Error())
		return
	}
	if body.Emoji == "" {
		writeJSONError(w, http.StatusBadRequest, "emoji is required")
		return
	}

	h.Store.AddReaction(messageID, user.ID, body.Emoji)

	reactions, err := h.Store.GetReactionsByMessage(messageID)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to get reactions")
		return
	}

	if h.Hub != nil {
		h.Store.CreateEvent(&store.Event{
			Kind:      "reaction_update",
			ChannelID: msg.ChannelID,
			Payload:   mustJSON(map[string]any{"message_id": messageID, "reactions": reactions}),
		})
		h.Hub.BroadcastEventToChannel(msg.ChannelID, "reaction_update", map[string]any{"message_id": messageID, "reactions": reactions})
	}

	writeJSONResponse(w, http.StatusOK, map[string]any{"ok": true, "reactions": reactions})
}

func (h *ReactionHandler) handleRemoveReaction(w http.ResponseWriter, r *http.Request) {
	user, ok := mustUser(w, r)
	if !ok {
		return
	}

	messageID := r.PathValue("messageId")
	msg, ok := h.canAccessMessage(user, messageID)
	if !ok {
		writeJSONError(w, http.StatusNotFound, "Channel not found")
		return
	}

	var body struct {
		Emoji string `json:"emoji"`
	}
	if err := readJSON(r, &body); err != nil {
		writeJSONError(w, http.StatusBadRequest, err.Error())
		return
	}
	if body.Emoji == "" {
		writeJSONError(w, http.StatusBadRequest, "emoji is required")
		return
	}

	h.Store.RemoveReaction(messageID, user.ID, body.Emoji)

	reactions, err := h.Store.GetReactionsByMessage(messageID)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to get reactions")
		return
	}

	if h.Hub != nil {
		h.Store.CreateEvent(&store.Event{
			Kind:      "reaction_update",
			ChannelID: msg.ChannelID,
			Payload:   mustJSON(map[string]any{"message_id": messageID, "reactions": reactions}),
		})
		h.Hub.BroadcastEventToChannel(msg.ChannelID, "reaction_update", map[string]any{"message_id": messageID, "reactions": reactions})
	}

	writeJSONResponse(w, http.StatusOK, map[string]any{"ok": true, "reactions": reactions})
}

func (h *ReactionHandler) handleGetReactions(w http.ResponseWriter, r *http.Request) {
	user, ok := mustUser(w, r)
	if !ok {
		return
	}

	messageID := r.PathValue("messageId")
	if _, ok := h.canAccessMessage(user, messageID); !ok {
		writeJSONError(w, http.StatusNotFound, "Channel not found")
		return
	}

	reactions, err := h.Store.GetReactionsByMessage(messageID)
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Failed to get reactions")
		return
	}

	writeJSONResponse(w, http.StatusOK, map[string]any{"reactions": reactions})
}
